### DISCUSSION TITLE
Welcome to h3 Discussions!

### DISCUSSION THREAD

**Discussion by isaacbrodsky** [uber/h3#416] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of the H3 community. We hope that you:
  * Ask questions about using H3.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

For bug reports and features requests we ask that you open a [Github issue](https://github.com/uber/h3/issues).

You can also ask questions by posting to the [H3 tag on StackOverflow](https://stackoverflow.com/questions/tagged/h3).

There is also an [H3 Slack workspace](https://join.slack.com/t/h3-core/shared_invite/zt-g6u5r1hf-W_~uVJmfeiWtMQuBGc1NNg) that we hope you will participate in.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by RichardVasquez** [comment]
Hi! I'm Richard Vasquez. I mainly lurk here, but I try to keep up to date on how H3 is progressing as I'm still working on my C# translation (not bridge) of H3. I need to find some free time to bring my version up to current.

I've been a little slow on that lately as I'm also using that C# implementation for a mobile game.

![image](https://user-images.githubusercontent.com/1706965/102559680-4bb65280-4085-11eb-997b-9dcae4847a2f.png)

**Comment by chninja** [comment]
Hi, 
Does anyone know if H3 uses any type of space filling curve like the S2 Geometry lib does ?
Thanks
Catherine

**Reply by nrabinowitz** [reply]
Not as such. H3 uses hierarchical subdivision - the world is divided into 122 base cells, each of those are divided into children, and so on. The index itself uses this hierarchical structure to identify a given cell, essentially as "base cell A, child B of A, child C of B, ...".

Detailed description of indexing: https://h3geo.org/docs/core-library/h3Indexing
Visual representation of indexing order: https://observablehq.com/@nrabinowitz/h3-indexing-order

**Reply by dfellis** [reply]
Well, within any particular base cell there's a predictable subdivision ordering, so in that sense it's like a space-filling curve, but the path between base cells doesn't follow that same pattern. (And really couldn't, due to the pentagons that would interfere with it).

Also that curve within a base cell is not contiguous within its own level like most space-filling curves are, so there wouldn't be the usual "next index in the list is the neighbor of the previous one if sharing the same parent" property.

**Reply by chninja** [reply]
@dfellis @nrabinowitz Thank you for the response and links. This helps.

**Comment by chninja** [comment]
Hi, I am using H3 as part of research and I am trying to figure out the complexity of some of the functions. I noticed that some of the functions were from files that were not in github repo like polygonalgos.h. Are some of these hidden somehow?

**Reply by nrabinowitz** [reply]
The `.h` files are in the repo - they're in `src/h3lib/include`: https://github.com/uber/h3/blob/master/src/h3lib/include/polygonAlgos.h

The polygon algorithms are a bit weird - they're defined as macros, because we have two different polygon data types (one based on arrays, one based on linked lists) to meet different memory management requirements. To make the same algorithms run over the two data types, we define them in macros and then define the data-type specific logic separately, composing macros that use the algo with one data type or another. This would be a lot easier in other languages, but was the best option we could find in C.

**Reply by chninja** [reply]
Thanks for the reply, I don't know why I couldn't see this.

**Comment by PandoraMC** [comment]
Is there acceleration support? like GPU.
Which is the better computer architecture for execution?

**Reply by isaacbrodsky** [reply]
This is something we're very interested in exploring, but I don't believe GPU acceleration for H3 has been developed. Contributions on this would be very welcome.

**Reply by BLTC-520** [reply]
Posting the same question after 4 years, is there any GPU acceleration for H3 has been developed?
